### PR TITLE
feat(expr): add `jsonb_build_array/object` function

### LINKED ISSUE #13016: support `jsonb_build_array` / `jsonb_build_object`
> binder & expr support of VARIADIC any is poor

Requested by a community user

### PR DESCRIPTION
**Pull request by wangrunji0408**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

resolve #13016 

This PR implements `jsonb_build_array` and `jsonb_build_object` function. But the `VARIADIC` keyword before arguments are not supported.

This PR also adds more data type support for `jsonb_agg` and `jsonb_object_agg`.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [x] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

Add `jsonb_build_array` and `jsonb_build_object` function:

```
jsonb_build_array ( VARIADIC "any" ) → jsonb
Builds a possibly-heterogeneously-typed JSON array out of a variadic argument list. Each argument is converted as per to_json or to_jsonb.
jsonb_build_array(1, 2, 'foo', 4, 5) → [1, 2, "foo", 4, 5]

jsonb_build_object ( VARIADIC "any" ) → jsonb
Builds a JSON object out of a variadic argument list. By convention, the argument list consists of alternating keys and values. Key arguments are coerced to text; value arguments are converted as per to_json or to_jsonb.
jsonb_build_object('foo', 1, 2, row(3,'bar')) → {"foo" : 1, "2" : {"f1":3,"f2":"bar"}}
```

Limitation: calling with `VARIADIC` keyword is not supported yet. e.g.

```
jsonb_build_array(VARIADIC array[1, 2] :: int[]) 
jsonb_build_object(VARIADIC array[1, 2] :: varchar[]) 
```

- `jsonb_agg` now accept any type except array and struct.
- `jsonb_object_agg` now accept any type except array and struct **as value**. keys are still limited to varchar.

### CODE REVIEW

**Review by KeXiangWang — APPROVED** [review]
Great refactoring. Thanks!

File: src/expr/impl/src/scalar/jsonb_build.rs [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+        .enumerate()
+    {
+        match key {
+            Some(ScalarRefImpl::List(_) | ScalarRefImpl::Struct(_) | ScalarRefImpl::Jsonb(_)) => {
+                return Err(ExprError::InvalidParam {
+                    name: "args",
+                    reason: "key value must be scalar, not array, composite, or json".into(),
+                })
+            }
+            Some(s) => builder.display(ToTextDisplay(s)),
```
Comment (xiangjinwu) [inline src/expr/impl/src/scalar/jsonb_build.rs:81]: caveat: `bool` does not use `ToText` and get [special](https://github.com/postgres/postgres/blob/REL_16_0/src/backend/utils/adt/jsonb.c#L797-L803) treatment.

RisingWave (this PR):
```
dev=> select jsonb_build_object(true, 2);
 jsonb_build_object 
--------------------
 {"t": 2}
(1 row)
```

PostgreSQL:
```
test=# select jsonb_build_object(true, 2);
 jsonb_build_object 
--------------------
 {"true": 2}
(1 row)

test=# select concat('', true);
 concat 
--------
 t
(1 row)

test=# select '' || true;
 ?column? 
----------
 true
(1 row)
```

**Review by xiangjinwu — APPROVED** [review]
rest lgtm

**Comment by fuyufjh** [comment]
Being asked by a user today. Please remember to tag as user-facing.

**Comment by wangrunji0408** [comment]
Thanks for reminding. I didn't click the checkbox initially, but clicked it later by editing. Looks like it didn't trigger the bot to automatically add the tag. (

**Comment by fuyufjh** [comment]
Indeed. The GitHub action seems to be broken... cc. @neverchanje 

https://github.com/risingwavelabs/risingwave/actions/workflows/auto-create-doc-issue-by-issue.yml

**Comment by fuyufjh** [comment]
Sorry my fault. @wangrunji0408 

- This PR just missed the v1.4 release, that's why it's not in the release note
- Tag `user-facing-changes` was not used since https://github.com/risingwavelabs/risingwave/pull/11591

**Pull request merged**